File: tourcount/src/main/java/com/wmstein/tourcount/widgets/EditIndividualWidget.java
package com.wmstein.tourcount.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.wmstein.tourcount.R;

/***************************************************************
 * EditIndividualWidget.java used by EditIndividualActivity.java
 * Created by wmstein for com.wmstein.tourcount on 2016-05-15.
 * Last edited on 2018-03-31
 */
public class EditIndividualWidget extends LinearLayout
{
    private final TextView widget_locality1; // locality
    private final EditText widget_locality2;
    private final TextView widget_zcoord1;  //height
    private final TextView widget_zcoord2;
    private final TextView widget_stadium1; // stadium
    private final TextView widget_stadium2;
    private final TextView widget_state1; // state_1-6
    private final EditText widget_state2;
    private final TextView widget_count1; // number of individuals
    private final EditText widget_count2;
    private final TextView widget_indivnote1; // note
    private final EditText widget_indivnote2;
    private final TextView widget_xcoord1; // x-coord
    private final TextView widget_xcoord2;
    private final TextView widget_ycoord1; // y-coord
    private final TextView widget_ycoord2;

    public EditIndividualWidget(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        inflater.inflate(R.layout.widget_edit_individual, this, true);
        widget_locality1 = findViewById(R.id.widgetLocality1); // Locality
        widget_locality2 = findViewById(R.id.widgetLocality2);
        widget_zcoord1 = findViewById(R.id.widgetZCoord1); // Height
        widget_zcoord2 = findViewById(R.id.widgetZCoord2);
        widget_stadium1 = findViewById(R.id.widgetStadium1); // Stadium
        widget_stadium2 = findViewById(R.id.widgetStadium2);
        widget_state1 = findViewById(R.id.widgetState1); // State_1-6
        widget_state2 = findViewById(R.id.widgetState2);
        widget_count1 = findViewById(R.id.widgetCount1); // number of individuals
        widget_count2 = findViewById(R.id.widgetCount2);
        widget_indivnote1 = findViewById(R.id.widgetIndivNote1); // Note
        widget_indivnote2 = findViewById(R.id.widgetIndivNote2);
        widget_xcoord1 = findViewById(R.id.widgetXCoord1); // X-Coord
        widget_xcoord2 = findViewById(R.id.widgetXCoord2);
        widget_ycoord1 = findViewById(R.id.widgetYCoord1); // Y-Coord
        widget_ycoord2 = findViewById(R.id.widgetYCoord2);
    }

    // Following the SETS
    // locality
    public void setWidgetLocality1(String title)
    {
        widget_locality1.setText(title);
    }

    // stadium
    public void setWidgetStadium1(String title)
    {
        widget_stadium1.setText(title);
    }

    // state
    public void setWidgetState1(String title)
    {
        widget_state1.setText(title);
    }

    // number of individuals
    public void setWidgetCount1(String title)
    {
        widget_count1.setText(title);
    }

    // note
    public void setWidgetIndivNote1(String title)
    {
        widget_indivnote1.setText(title);
    }

    // x-coord
    public void setWidgetXCoord1(String title)
    {
        widget_xcoord1.setText(title);
    }

    public void setWidgetXCoord2(String name)
    {
        widget_xcoord2.setText(name);
    }

    // y-coord
    public void setWidgetYCoord1(String title)
    {
        widget_ycoord1.setText(title);
    }

    public void setWidgetYCoord2(String name)
    {
        widget_ycoord2.setText(name);
    }

    // z-coord
    public void setWidgetZCoord1(String title)
    {
        widget_zcoord1.setText(title);
    }

    public void setWidgetZCoord2(String name)
    {
        widget_zcoord2.setText(name);
    }

    // following the GETS
    // get locality
    public String getWidgetLocality2()
    {
        return widget_locality2.getText().toString();
    }

    public void setWidgetLocality2(String name)
    {
        widget_locality2.setText(String.valueOf(name));
    }

    // get stadium with plausi
    public String getWidgetStadium2()
    {
        return widget_stadium2.getText().toString();
    }

    public void setWidgetStadium2(String name)
    {
        widget_stadium2.setText(name);
    }

    // get state with plausi
    public int getWidgetState2()
    {
        String text = widget_state2.getText().toString();
        String regEx = "^[0-9]*$";
        if (text.equals(""))
            return 0;
        else if (!text.trim().matches(regEx))
            return 100;
        else
        {
            try
            {
                return Integer.parseInt(text.replaceAll("[\\D]",""));
            } catch (NumberFormatException nfe)
            {
                return 0;
            }
        }
    }

    public void setWidgetState2(int name)
    {
        widget_state2.setText(String.valueOf(name));
    }

    // get number of individuals
    public int getWidgetCount2()
    {
        String text = widget_count2.getText().toString();
            try
            {
                // value >= 0
                return Integer.parseInt(text.replaceAll("[\\D]",""));
            } catch (NumberFormatException nfe)
            {
                return -1; // count < 0 or text has no digit
            }
    }

    public void setWidgetCount2(int name)
    {
        widget_count2.setText(String.valueOf(name));
    }

    // get PLZ with plausi
    public String getWidgetIndivNote2()
    {
        return widget_indivnote2.getText().toString();
    }

    public void setWidgetIndivNote2(String name)
    {
        widget_indivnote2.setText(name);
    }

}

File: tourcount/src/main/java/com/wmstein/tourcount/widgets/EditIndividualWidget.kt
package com.wmstein.tourcount.widgets

import android.annotation.SuppressLint
import android.content.Context
import android.text.Editable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView

import com.wmstein.tourcount.AutoFitEditText
import com.wmstein.tourcount.R

import java.util.Objects

/*************************************************************
 * EditIndividualWidget.kt used by EditIndividualActivity.kt
 *
 * Created by wmstein for com.wmstein.tourcount on 2016-05-15.
 * Last edited in Java on 2022-03-26,
 * converted to Kotlin on 2023-07-09,
 * last edited on 2026-08-13
 */
class EditIndividualWidget(context: Context, attrs: AttributeSet?) : LinearLayout(context, attrs) {
    // locality
    private val widgetLoc1: TextView
    private val widgetLoc2: EditText

    // stadium
    private val widgetStad1: TextView
    private val widgetStad2: TextView

    // state_1-6
    private val widgetStat1: TextView
    private val widgetStat2: EditText

    // number of individuals
    private val widgetCnt1: TextView
    private val widgetCnt2: AutoFitEditText

    // x-coord
    private val widgetXcoord1: TextView
    private val widgetXcoord2: TextView

    // y-coord
    private val widgetYcoord1: TextView
    private val widgetYcoord2: TextView

    // height
    private val widgetZcoord1: TextView
    private val widgetZcoord2: TextView

    // notes
    private val widgetIndNote1: TextView
    private val widgetIndNote2: AutoCompleteTextView

    private val suggestions = ArrayList(listOf(*resources.getStringArray(R.array.indiv_notes_options)))
    private val indivNotesAdapter = ArrayAdapter(context, android.R.layout.select_dialog_item,
        suggestions)
    private var selectedValue = ""

    val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    init {
        Objects.requireNonNull(inflater)
            .inflate(R.layout.widget_edit_individual, this, true)

        widgetLoc1 = findViewById(R.id.widget_Locality1) // Locality
        widgetLoc2 = findViewById(R.id.widget_Locality2)
        widgetZcoord1 = findViewById(R.id.widget_ZCoord1) // Height
        widgetZcoord2 = findViewById(R.id.widget_ZCoord2)

        widgetXcoord1 = findViewById(R.id.widget_XCoord1) // Latitude
        widgetXcoord2 = findViewById(R.id.widget_XCoord2)
        widgetYcoord1 = findViewById(R.id.widget_YCoord1) // Longitude
        widgetYcoord2 = findViewById(R.id.widget_YCoord2)

        widgetStad1 = findViewById(R.id.widget_Stadium1) // Phase
        widgetStad2 = findViewById(R.id.widget_Stadium2)
        widgetStat1 = findViewById(R.id.widget_State1) // State_1-6
        widgetStat2 = findViewById(R.id.widget_State2)
        widgetCnt1 = findViewById(R.id.widget_Count1) // Number of individuals
        widgetCnt2 = findViewById(R.id.widget_Count2)

        widgetIndNote1 = findViewById(R.id.widget_IndivNote1) // Notes head
        widgetIndNote2 = findViewById(R.id.widget_IndivNote2) // Notes text
        widgetIndNote2.setAdapter(indivNotesAdapter)
        widgetIndNote2.setOnFocusChangeListener {
                _, hasFocus ->
            if (hasFocus)
                widgetIndNote2.showDropDown()
        }
        widgetIndNote2.setOnItemClickListener {
                parent, _, position, _ ->
            selectedValue = parent.getItemAtPosition(position).toString()
        }
        setIndivNotes(selectedValue)
    }

    // Following the SETS
    // Locality
    fun setWidgetLocality1(title: String) {
        widgetLoc1.text = title
    }

    // Height
    fun setWidgetZCoord1(title: String) {
        widgetZcoord1.text = title
    }

    fun setWidgetZCoord2(name: String) {
        widgetZcoord2.text = name
    }

    // Latitude
    fun setWidgetXCoord1(title: String) {
        widgetXcoord1.text = title
    }

    fun setWidgetXCoord2(name: String) {
        widgetXcoord2.text = name
    }

    // Longitude
    fun setWidgetYCoord1(title: String) {
        widgetYcoord1.text = title
    }

    fun setWidgetYCoord2(name: String) {
        widgetYcoord2.text = name
    }

    // Phase
    fun setWidgetStadium1(title: String) {
        widgetStad1.text = title
    }

    // Status
    fun setWidgetState1(title: String) {
        widgetStat1.text = title
    }

    // Number of individuals
    fun setWidgetCount1(title: String) {
        widgetCnt1.text = title
    }

    // Notes
    fun setWidgetIndivNote1(title: String) {
        widgetIndNote1.text = title
    }

    // Following the GETS
    // get locality
    var widgetLocality2: String
        get() = widgetLoc2.text.toString()
        set(name) {
            widgetLoc2.setText(name)
        }

    // get stadium with plausi
    var widgetStadium2: String
        get() = widgetStad2.text.toString()
        set(name) {
            widgetStad2.text = name
        }

    fun widgetState1(enabled: Boolean) {
        if (enabled) widgetStat1.visibility = VISIBLE else widgetStat1.visibility = INVISIBLE
    }

    // get state number with plausi
    val widgetState2: String
        get() {
            var text = widgetStat2.text.toString()
            if (text == "-") text = "0"
            val regEx = Regex("^[0-9]*$")
            return if (text == "") "0" else if (!text.trim { it <= ' ' }
                    .matches(regEx)) "100" else {
                try {
                    text.replace("\\D".toRegex(), "")
                } catch (_: NumberFormatException) {
                    "0"
                }
            }
        }

    fun widgetState2(enabled: Boolean) {
        if (enabled) widgetStat2.visibility = VISIBLE else widgetStat2.visibility = INVISIBLE
    }

    fun setWidgetState2(name: String) {
        widgetStat2.setText(name)
    }

    // get number of individuals
    var widgetCount2: Int
        get() {
            val text = widgetCnt2.text.toString()
            return try {
                // value >= 0
                text.replace("\\D".toRegex(), "").toInt()
            } catch (_: NumberFormatException) {
                -1 // count < 0 or text has no digit
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            widgetCnt2.setText(name.toString())
        }

    // Get selected text for notes
    fun setIndivNotes(note: String) {
        widgetIndNote2.text = Editable.Factory.getInstance().newEditable(note)
    }

    // Get notes of individual
    var widgetIndivNote2: String
        get() = widgetIndNote2.text.toString()
        set(name) {
            widgetIndNote2.setText(name)
        }

}
